### PR TITLE
Libflutter engine rework

### PR DESCRIPTION
**Pull request by pieter-scholtz**
Adding support profiling mode and ability to look for libflutter_engine.so in the asset bundle.

### CODE REVIEW

File: include/flutter-pi.h [hunk 1]
```diff
@@ -313,7 +313,7 @@ struct frame {
 struct compositor;
 
 enum flutter_runtime_mode {
-	kDebug, kRelease
+	kDebug, kProfile, kRelease
```
Comment (ardera) [inline include/flutter-pi.h:316]: I just noticed, you added `kProfile` as a runtime mode and `--profile` to the usage, but you never added anything to the commandline args parsing that'll recognize that argument and also you didn't add anything to the flutter engine loading that'll load `libflutter_engine.so.profile` if profile mode was chosen.

File: src/flutter-pi.c [hunk 2]
```diff
@@ -1680,7 +1686,20 @@ static int init_application(void) {
...
 	libflutter_engine_handle = NULL;
-	if (flutterpi.flutter.runtime_mode == kRelease) {
+
+    libflutter_engine_handle = dlopen(flutterpi.flutter.libflutter_engine_path, RTLD_LOCAL | RTLD_NOW);
+
+    if (libflutter_engine_handle == NULL)
+    {
+        LOG_FLUTTERPI_ERROR(
+            "[flutter-pi] Warning: Could not load libflutter_engine.so inside the asset bundle : "
+            "%s. Trying to open libflutter_engine.so.${runtimeMode} ...\n",
```
Comment (ardera) [inline src/flutter-pi.c:1696]: `${}` string interpolation came like 50 years after C was invented, so it won't work here I'm afraid ;)

```suggestion
            "%s. Trying to open libflutter_engine.so.%s ...\n",
            dlerror(), flutterpi.flutter.runtime_mode == kDebug ? "debug" : flutterpi.flutter.runtime_mode == kProfile ? "profile" : "debug"
```

**Review by ardera — COMMENTED** [review]
if you fix the things I commented on it LGTM

File: include/flutter-pi.h (hunk 1 above)
Comment (pieter-scholtz) [reply include/flutter-pi.h:218]: where do I find the flutter engine loading , as well as commandline args parsing ?

Comment (ardera) [reply include/flutter-pi.h:218]: flutter engine loading is here:
https://github.com/ardera/flutter-pi/blob/e306d1b9ce8766c82f2415b1f4f7ae6ae91f90e3/src/flutter-pi.c#L1816

cmdline arg parsing is here:
https://github.com/ardera/flutter-pi/blob/e306d1b9ce8766c82f2415b1f4f7ae6ae91f90e3/src/flutter-pi.c#L2300

basically, just search for `kRelease` or `kDebug` in flutter-pi.c and ensure for every occurence that you handle `kProfile` there correctly

File: src/flutter-pi.c [hunk 3]
```diff
@@ -1810,25 +1816,63 @@ static int init_application(void) {
...
+        if (libflutter_engine_handle == NULL) {
+            LOG_ERROR("Warning: Could not load libflutter_engine.so.profile: %s Trying to open libflutter_engine.so.debug", dlerror());
+            printf("\n");
         }
     } else if (flutterpi.flutter.runtime_mode == kDebug) {
         libflutter_engine_handle = dlopen("libflutter_engine.so.debug", RTLD_LOCAL | RTLD_NOW);
         if (libflutter_engine_handle == NULL) {
-            LOG_ERROR("[flutter-pi] Warning: Could not load libflutter_engine.so.debug: %s. Trying to open libflutter_engine.so...\n", dlerror());
+            LOG_ERROR("Warning: Could not load libflutter_engine.so.debug: %s Trying to open libflutter_engine.so ", dlerror());
+            printf("\n");
```
Comment (ardera) [inline src/flutter-pi.c:1866]: Why did you put the `\n` into a separate printf here and above?
The printf will go to stdout, LOG_ERROR will go to stderr

File: src/flutter-pi.c [hunk 4]
```diff
@@ -1810,25 +1816,63 @@ static int init_application(void) {
...
+
     libflutter_engine_handle = NULL;
+
+    libflutter_engine_handle = dlopen(libflutter_engine_path, RTLD_LOCAL | RTLD_NOW);
+
+    if (libflutter_engine_handle == NULL)
+    {   
+
+        
+        printf("[flutter-pi] Warning: Could not load libflutter_engine.so inside the asset bundle : Trying to open libflutter_engine.so.");
```
Comment (ardera) [inline src/flutter-pi.c:1832]: Can you merge these into one `LOG_ERROR`?

File: src/flutter-pi.c [hunk 5]
```diff
@@ -1810,26 +1816,49 @@ static int init_application(void) {
...
 
+    asprintf(&libflutter_engine_path, "%s/libflutter_engine.so", flutterpi.flutter.asset_bundle_path);
+
     libflutter_engine_handle = NULL;
+
+    libflutter_engine_handle = dlopen(libflutter_engine_path, RTLD_LOCAL | RTLD_NOW);
+
+    if (libflutter_engine_handle == NULL)
+    {   
+        LOG_ERROR("Warning: Could not load libflutter_engine.so inside the asset bundle : Trying to open libflutter_engine.so.%s \n",flutterpi.flutter.runtime_mode == kDebug ? "debug" :
```
Comment (ardera) [inline src/flutter-pi.c:1830]:
```suggestion
        LOG_ERROR("Warning: Could not load libflutter_engine.so from the asset bundle: Trying to open libflutter_engine.so.%s \n",flutterpi.flutter.runtime_mode == kDebug ? "debug" :
```

File: src/flutter-pi.c [hunk 6]
```diff
@@ -1810,26 +1816,49 @@ static int init_application(void) {
...
     if (flutterpi.flutter.runtime_mode == kRelease) {
         libflutter_engine_handle = dlopen("libflutter_engine.so.release", RTLD_LOCAL | RTLD_NOW);
         if (libflutter_engine_handle == NULL) {
-            LOG_ERROR("[flutter-pi] Warning: Could not load libflutter_engine.so.release: %s. Trying to open libflutter_engine.so...\n", dlerror());
+            LOG_ERROR("Warning: Could not load libflutter_engine.so.release: Trying to open libflutter_engine.so.profile \n");
+        }
+    } else if (flutterpi.flutter.runtime_mode == kProfile) {
+        libflutter_engine_handle = dlopen("libflutter_engine.so.profile", RTLD_LOCAL | RTLD_NOW);
+        if (libflutter_engine_handle == NULL) {
+            LOG_ERROR("Warning: Could not load libflutter_engine.so.profile: Trying to open libflutter_engine.so.debug \n");
```
Comment (ardera) [inline src/flutter-pi.c:1847]:
```suggestion
            LOG_ERROR("Warning: Could not load libflutter_engine.so.profile. dlopen: %s. Trying to open libflutter_engine.so...\n");
```

File: src/flutter-pi.c [hunk 7]
```diff
@@ -1810,26 +1816,49 @@ static int init_application(void) {
...
         }
     } else if (flutterpi.flutter.runtime_mode == kDebug) {
         libflutter_engine_handle = dlopen("libflutter_engine.so.debug", RTLD_LOCAL | RTLD_NOW);
         if (libflutter_engine_handle == NULL) {
-            LOG_ERROR("[flutter-pi] Warning: Could not load libflutter_engine.so.debug: %s. Trying to open libflutter_engine.so...\n", dlerror());
+            LOG_ERROR("Warning: Could not load libflutter_engine.so.debug: Trying to open libflutter_engine.so \n");
```
Comment (ardera) [inline src/flutter-pi.c:1852]:
```suggestion
            LOG_ERROR("Warning: Could not load libflutter_engine.so.debug. dlopen: %s. Trying to open libflutter_engine.so...\n", dlerror());
```

File: src/flutter-pi.c [hunk 8]
```diff
@@ -1810,26 +1816,49 @@ static int init_application(void) {
...
         }
     }
+    }
 
     if (libflutter_engine_handle == NULL) {
         libflutter_engine_handle = dlopen("libflutter_engine.so", RTLD_LOCAL | RTLD_NOW);
         if (libflutter_engine_handle == NULL) {
-            LOG_ERROR("Could not load libflutter_engine.so. dlopen: %s", dlerror());
-            LOG_ERROR("Could not find a fitting libflutter_engine.\n");
+            LOG_ERROR("Could not load libflutter_engine.so dlerror: %s \n", dlerror());
```
Comment (ardera) [inline src/flutter-pi.c:1860]:
```suggestion
            LOG_ERROR("Could not load libflutter_engine.so. dlopen: %s\n", dlerror());
```

File: src/flutter-pi.c [hunk 9]
```diff
@@ -1810,26 +1816,49 @@ static int init_application(void) {
...
+            LOG_ERROR("Could not find a fitting libflutter_engine.so , make sure you installed the engine binaries.\n");
```
Comment (ardera) [inline src/flutter-pi.c:1861]:
```suggestion
            LOG_ERROR("Could not find a fitting libflutter_engine.so, make sure you've installed the engine binaries.\n");
```

File: src/flutter-pi.c [hunk 10]
```diff
@@ -2275,7 +2304,7 @@ static bool setup_paths(void) {
             LOG_ERROR("Could not find \"kernel.blob\" file inside \"%s\", which is required for debug mode.\n", flutterpi.flutter.asset_bundle_path);
             return false;
         }
-    } else if (flutterpi.flutter.runtime_mode == kRelease) {
+    } else if ((flutterpi.flutter.runtime_mode == kRelease)||(flutterpi.flutter.runtime_mode == kProfile)) {
```
Comment (ardera) [inline src/flutter-pi.c:2307]:
```suggestion
    } else if ((flutterpi.flutter.runtime_mode == kRelease) || (flutterpi.flutter.runtime_mode == kProfile)) {
```

File: src/flutter-pi.c [hunk 11]
```diff
@@ -1810,26 +1816,49 @@ static int init_application(void) {
...
+
+    free(libflutter_engine_path);
+
+    if (libflutter_engine_handle == NULL)
+    {
     if (flutterpi.flutter.runtime_mode == kRelease) {
         libflutter_engine_handle = dlopen("libflutter_engine.so.release", RTLD_LOCAL | RTLD_NOW);
         if (libflutter_engine_handle == NULL) {
-            LOG_ERROR("[flutter-pi] Warning: Could not load libflutter_engine.so.release: %s. Trying to open libflutter_engine.so...\n", dlerror());
+            LOG_ERROR("Warning: Could not load libflutter_engine.so.release: Trying to open libflutter_engine.so.profile \n");
```
[the remark this answers is not in the document]
Comment (ardera) [reply src/flutter-pi.c:1842]: I think that's the last one

File: src/flutter-pi.c [hunk 12]
```diff
@@ -1810,26 +1816,49 @@ static int init_application(void) {
...
     int ok;
 
+    asprintf(&libflutter_engine_path, "%s/libflutter_engine.so", flutterpi.flutter.asset_bundle_path);
+
     libflutter_engine_handle = NULL;
+
+    libflutter_engine_handle = dlopen(libflutter_engine_path, RTLD_LOCAL | RTLD_NOW);
+
+    if (libflutter_engine_handle == NULL)
+    {
```
Comment (ardera) [inline src/flutter-pi.c:1829]:
```suggestion
    if (libflutter_engine_handle == NULL) {   
```

File: src/flutter-pi.c [hunk 13]
```diff
@@ -1810,26 +1816,49 @@ static int init_application(void) {
...
+    {   
+        LOG_ERROR("Warning: Could not load libflutter_engine.so from the asset bundle: Trying to open libflutter_engine.so.%s\n", flutterpi.flutter.runtime_mode == kDebug ? "debug" :
+            flutterpi.flutter.runtime_mode == kProfile ? "profile" :
+            "release");
+    }
+
+    free(libflutter_engine_path);
+
+    if (libflutter_engine_handle == NULL)
+    {
```
Comment (ardera) [inline src/flutter-pi.c:1838]:
```suggestion
    if (libflutter_engine_handle == NULL) {
```

**Comment by ardera** [comment]
Just tried running, I get an error when trying to run in profile mode.

In line https://github.com/ardera/flutter-pi/blob/e2d185370dc41ace05864d849666b187c55355ce/src/flutter-pi.c#L2013, you need to check if `runtime_mode == kProfile` too, otherwise the `app.so` is not loaded.

**Review by ardera — APPROVED** [review]

**Pull request merged**